pkg/query/api/v1.go

```diff
@@ -269,6 +270,21 @@ func (api *API) parseReplicaLabelsParam(r *http.Request) (replicaLabels []string
 	return replicaLabels, nil
 }
 
+func (api *API) parseStoreAllowListParam(r *http.Request) (storeAllowList []string, _ *ApiError) {
+	const storeAllowListParam = "storeAllowList[]"
```

Comment:
Now when I think of it I wonder if something like matchers would be quite neat to have.
Because when we add allowlist we will soon have request "can I have denylist"? ;p

I mean something like this API have: https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers

For example:
 
```
&storeMatch[]={__address__=~"1\.2\.3\.\d*"}
```
This will give us quite amazing flexibility as in future we can add more "labels" to maybe filter by source, health, or any other metadata ... what do you think? (:

Reply:
Maybe it could rely on the same system as the external labels match:
- When a store is registered we add a label for `__thanos_store_address__` containing the address (and maybe a `__thanos_store_health__` as well)
- Then the user add the label in its regular query and we match it using the regular system
- Before we dispatch the query to the other store, we drop all the `__thanos_store_.*` labels

Pro:
- easy to implement
- easy to use

Con:
- magic is done is the background
- we reserve some labels

What do you think?

Reply:
Yea I was thinking about this as well, but this sounds scary. I think this one will need more thoughts. Let me propagate this question among maintainers and devs. I think the matcher idea is less intrusive.

We could in theory add those labels only in some `--debug` mode :thinking:

Reply:
We definitely cannot inject them as normal flow as it will surprise Querier clients. It's important to remember that users already build alerts etc so data consistency is the key

Reply:
wonder what e.g @brian-brazil would say about our ideas =D

Reply:
I see no problems here with storeMatch - though I wonder if you really need regexes if this is just for ad-hoc debugging.

Tying this to general label semantics sounds like it's over-complicating things, and would likely cause problems down time line (plus is technically breaking as `__thanos_store_address__` is a valid label value that could come from Prometheus).

Reply:
Alright, let's go for `storeMatch` then! 👍

Reply:
Yes. Although I really like the fact that you could aggregate, join etc all metadata. ;p

Reply:
I updated the PR. Much cleaner to use matcher instead of a regular allow list!